mirrord-layer/src/pod_api.rs

```diff
@@ -46,56 +49,24 @@ struct RuntimeData {
     container_runtime: String,
     node_name: String,
     socket_path: String,
+    pod_name: String,
+    container_name: String,
 }
 
 impl RuntimeData {
-    async fn from_k8s(
-        client: Client,
-        pod_name: &str,
-        pod_namespace: &str,
-        container_name: &Option<String>,
-    ) -> Result<Self> {
-        let pods_api: Api<Pod> = Api::namespaced(client, pod_namespace);
-        let pod = pods_api.get(pod_name).await?;
-        let node_name = &pod.spec.unwrap().node_name;
-        let container_statuses = &pod.status.unwrap().container_statuses.unwrap();
-        let container_info = if let Some(container_name) = container_name {
-            &container_statuses
-                .iter()
-                .find(|&status| &status.name == container_name)
-                .ok_or_else(|| {
-                    LayerError::ContainerNotFound(
-                        container_name.clone(),
-                        pod_namespace.to_string(),
-                        pod_name.to_string(),
-                    )
-                })?
-                .container_id
-        } else {
-            info!("No container name specified, defaulting to first container found");
-            &container_statuses.first().unwrap().container_id
-        };
+    async fn from_k8s(client: Client, target: &str, target_namespace: &str) -> Result<Self> {
+        let target = target.parse::<Target>()?;
 
-        let container_info = container_info
+        let response = target.container_info(&client, target_namespace).await?;
+        let container_info = response
+            .container_info
             .as_ref()
-            .unwrap()
-            .split("://")
-            .collect::<Vec<&str>>();
-
-        let (container_runtime, socket_path) = match container_info.first() {
-            Some(&"docker") => ("docker", "/var/run/docker.sock"),
-            Some(&"containerd") => ("containerd", "/run/containerd/containerd.sock"),
-            _ => panic!("unsupported container runtime"),
-        };
-
-        let container_id = container_info.last().unwrap();
+            .ok_or_else(|| {
+                LayerError::ContainerNotFound(target_namespace.to_string(), target.clone())
+            })?
+            .parse::<ContainerRuntime>()?;
 
-        Ok(RuntimeData {
-            container_id: container_id.to_string(),
-            container_runtime: container_runtime.to_string(),
-            node_name: node_name.as_ref().unwrap().to_string(),
-            socket_path: socket_path.to_string(),
-        })
+        target.merge(response, container_info)
```

Comment:
Not sure what you're going for here. `response_data` and `container_info` both originate in `target`. Why is it the job of `RuntimeData` to merge them? Why doesn't `target` just return RuntimeData?
Taking this a bit further, is `impl RuntimeData` even necessary anymore, or can `from_k8s` be a method of `target`?

Reply:
- i was merging it in target, because I don't have to pass more data in function arguments for the custom errors and can be referred by self
- yup, I don't think we need it. i was just trying to preserve what was done previously